packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/shared/SubscriptionExtensionsTests.kt

```diff
@@ -0,0 +1,343 @@
+    }
+
+    // Check that subscribing twice to a query will result in the same subscription
+    @Test
+    fun realmQuery_subscribe_anonymousTwice() = runBlocking {
+        val subs = realm.subscriptions
+        assertEquals(0, subs.size)
+        realm.query<FlexParentObject>().subscribe()
+        realm.query<FlexParentObject>().subscribe()
+        assertEquals(SubscriptionSetState.COMPLETE, subs.state)
+        assertEquals(1, subs.size)
+        val sub: Subscription = subs.first()
+        assertNull(sub.name)
+        assertEquals("TRUEPREDICATE ", sub.queryDescription)
+        assertEquals(FlexParentObject::class.simpleName, sub.objectType)
+    }
+
+    // Check that anonymous RealmQuery and RealmResults .subscribe calls result in the same sub.
+    @Test
+    fun anonymousSubscriptionsOverlap() = runBlocking {
+        val subs = realm.subscriptions
+        assertEquals(0, subs.size)
+        realm.query<FlexParentObject>().subscribe()
+        realm.query<FlexParentObject>().find().subscribe()
+        assertEquals(SubscriptionSetState.COMPLETE, subs.state)
+        assertEquals(1, subs.size)
+        val sub: Subscription = subs.first()
+        assertNull(sub.name)
+        assertEquals("TRUEPREDICATE ", sub.queryDescription)
+        assertEquals(FlexParentObject::class.simpleName, sub.objectType)
+    }
+
+    // Verify that the realm query doesn't run against a frozen version previous to the Realm
+    // being updated from `waitForSynchronization`.
+    @Test
+    fun realmQuery_subscribe_queryResultIsLatestVersion() = runBlocking {
+        // Write data to a server Realm
+        val section = Random.nextInt()
+        val (email, password) = TestHelper.randomEmail() to "password1234"
+        val user1 = app.createUserAndLogIn(email, password)
+        val config = SyncConfiguration.Builder(
+            user1,
+            schema = setOf(FlexParentObject::class, FlexChildObject::class, FlexEmbeddedObject::class)
+        ).initialSubscriptions { realm: Realm ->
+            realm.query<FlexParentObject>("section = $0", section).subscribe()
+        }.build()
+
+        Realm.open(config).use { realmFromAnotherDevice ->
+            realmFromAnotherDevice.writeBlocking {
+                copyToRealm(FlexParentObject(section))
+            }
+            realmFromAnotherDevice.syncSession.uploadAllLocalChanges(30.seconds)
+        }
+
+        // Data still hasn't reached this device
+        assertEquals(0, realm.query<FlexParentObject>().count().find())
+        // Check that subscribing to a query, will run the query on the data downloaded from
+        // the server and not just local data, due to WaitForSync.FIRST_TIME being the default.
+        val result = realm.query<FlexParentObject>("section = $0", section).subscribe()
+        assertEquals(1, result.size)
+        assertEquals(section, result.first().section)
+    }
+
+    @Test
+    fun realmQuery_subscribe_waitFirstTime() = runBlocking<Unit> {
```

Comment:
There aren't really any logic testing that it is actually `waitFirstTime`'ing here are there 🤔 It looks like that it is the above test that actually tests that, at least data wise. Are we testing the effect on the subscription set? ... and shouldn't it then show that we are not `COMPLETE` after updating an existing subscription.

Reply:
This test calls `subscribe()` with has an implicit value of `waitFirstTime`, so I think the test is correct?

I.e. the subscription set should be COMPLETE with this setting, which is what the test asserts?

I guess it could be more visible if I made the mode explicit, but I wanted to test the default value on purpose.

Reply:
I see that we verify that it is `COMPLETE` due to `waitFirstTime`, but my point was that we should then also test that if we update the subscription then it is not `COMPLETE` immediately after the update, since it is only waiting the first time. Does that make sense?

Reply:
Ah, yeah, good point 👍